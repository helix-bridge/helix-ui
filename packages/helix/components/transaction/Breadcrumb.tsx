import { Breadcrumb } from 'antd';
import BreadcrumbItem from 'antd/lib/breadcrumb/BreadcrumbItem';
import { useTranslation } from 'next-i18next';

export function IBreadcrumb({ txHash }: { txHash: string }) {
  const { t } = useTranslation();

  return (
    <Breadcrumb
      className="whitespace-nowrap flex items-center overflow-hidden overflow-ellipsis"
      style={{ marginTop: '1.5rem' }}
    >
      <BreadcrumbItem>{t('Explorer')}</BreadcrumbItem>
      <BreadcrumbItem>{t('Transaction')}</BreadcrumbItem>
      <BreadcrumbItem>
        <span className="w-32 md:w-72 lg:w-96 truncate">{txHash}</span>
      </BreadcrumbItem>
    </Breadcrumb>
  );
}
