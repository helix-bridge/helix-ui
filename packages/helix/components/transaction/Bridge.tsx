import { useRouter } from 'next/router';
import { useMemo } from 'react';
import { Logo } from 'shared/components/widget/Logo';
import { Network } from 'shared/model';
import { getBridge } from 'shared/utils/bridge';
import { getChainConfig, getDisplayName } from 'shared/utils/network';

export function Bridge() {
  const router = useRouter();

  const [departure, arrival] = useMemo(
    () => [getChainConfig(router.query.from as Network), getChainConfig(router.query.to as Network)],
    [router.query]
  );

  const bridge = getBridge([departure, arrival]);

  return (
    <div>
      <div
        className="flex justify-between items-center gap-4 bg-gray-200 dark:bg-antDark bg-opacity-25"
        style={{ borderRadius: 40 }}
      >
        <div className="p-2 pr-0 flex items-center">
          <Logo chain={departure} width={40} height={40} className="w-5 md:w-10" />
        </div>

        <div
          className="flex items-center justify-center self-stretch px-4 md:px-8 bg-gray-300 bg-opacity-20 dark:bg-blue-900 dark:bg-opacity-25 relative"
          style={{
            clipPath: 'polygon(85% 0%, 100% 50%, 85% 100%, 0% 100%, 15% 50%, 0% 0%)',
          }}
        >
          <img src={`/image/bridges/${bridge.category}.png`} className="w-10 md:w-28" />
        </div>

        <div className="p-2 pl-0 flex items-center">
          <Logo chain={arrival} width={40} height={40} className="w-5 md:w-10" />
        </div>
      </div>

      <div className="flex justify-between text-xs text-white capitalize mt-1 px-3 whitespace-nowrap">
        {[departure, arrival].map((item) => (
          <div key={item.name}>{getDisplayName(item)}</div>
        ))}
      </div>
    </div>
  );
}
