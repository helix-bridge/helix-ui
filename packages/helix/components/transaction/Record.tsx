import { Tooltip } from 'antd';
import { formatDistance, fromUnixTime } from 'date-fns';
import { CrossChainState } from 'shared/components/widget/CrossChainStatus';
import { EllipsisMiddle } from 'shared/components/widget/EllipsisMiddle';
import { HelixHistoryRecord } from 'shared/model';
import { isDVM2Substrate } from 'shared/utils/bridge';
import { fromWei, prettyNumber, revertAccount } from 'shared/utils/helper';
import { getChainConfig, getDisplayName, toVertices } from 'shared/utils/network';

// eslint-disable-next-line complexity
export function Record({ record }: { record: HelixHistoryRecord }) {
  const { fromChain, sender, recipient, toChain } = record;
  const departure = toVertices(fromChain);
  const arrival = toVertices(toChain);
  const fromConfig = getChainConfig(departure);
  const toConfig = getChainConfig(arrival);
  const fromAccount = revertAccount(sender, fromConfig);
  const toAccount = revertAccount(recipient, toConfig);

  const amount = fromWei({ value: record.amount, decimals: isDVM2Substrate(departure, arrival) ? 18 : 9 }, (val) =>
    prettyNumber(val, { ignoreZeroDecimal: true })
  );

  const tokenName = !record.token.startsWith('0x')
    ? record.token
    : `${departure.mode === 'dvm' ? 'x' : ''}${fromConfig?.isTest ? 'O' : ''}RING`;

  return (
    <>
      <span className="justify-self-start ellipse-two-lines">
        {formatDistance(fromUnixTime(record.startTime), new Date(new Date().toUTCString()), {
          includeSeconds: true,
          addSuffix: true,
        })}
      </span>

      <div className="flex flex-col col-span-3 overflow-hidden pl-4 pr-2">
        <Tooltip title={fromAccount}>
          <span className="capitalize">{getDisplayName(fromConfig)}</span>
        </Tooltip>
        <EllipsisMiddle deviation={5} isGrow>
          {fromAccount}
        </EllipsisMiddle>
      </div>

      <div className="flex flex-col col-span-3 overflow-hidden pl-2 pr-4">
        <Tooltip title={toAccount}>
          <span className="capitalize">{getDisplayName(toConfig)}</span>
        </Tooltip>
        <EllipsisMiddle deviation={5} isGrow>
          {toAccount}
        </EllipsisMiddle>
      </div>

      <span>{tokenName}</span>

      <Tooltip title={amount}>
        <span className="justify-self-center max-w-full truncate">
          {prettyNumber(amount, { decimal: 3, ignoreZeroDecimal: true })}
        </span>
      </Tooltip>

      <span className="justify-self-center">
        {fromWei({ value: record.fee, decimals: departure.mode === 'dvm' ? 18 : 9 })}
      </span>

      <span className="justify-self-center capitalize">{record.bridge}</span>
      <CrossChainState value={record.result} />
    </>
  );
}
