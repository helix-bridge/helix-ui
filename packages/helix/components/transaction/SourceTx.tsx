import { useTranslation } from 'next-i18next';
import { useRouter } from 'next/router';
import { SubscanLink } from 'shared/components/widget/SubscanLink';
import { TextWithCopy } from 'shared/components/widget/TextWithCopy';
import { Network } from 'shared/model';
import { TransferDescription } from './TransferDescription';

export function SourceTx({ hash }: { hash: string | undefined }) {
  const { t } = useTranslation();
  const router = useRouter();

  return (
    <TransferDescription
      title={t('Source Tx Hash')}
      tip={t('Unique character string (TxID) assigned to every verified transaction on the Source Chain.')}
    >
      {hash && (
        <SubscanLink
          network={router.query.from as Network}
          txHash={hash}
          className="hover:opacity-80 transition-opacity duration-200"
        >
          <TextWithCopy underline>{hash}</TextWithCopy>
        </SubscanLink>
      )}
    </TransferDescription>
  );
}
