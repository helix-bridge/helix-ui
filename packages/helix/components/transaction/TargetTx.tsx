import { Progress, Typography } from 'antd';
import { useTranslation } from 'next-i18next';
import { useRouter } from 'next/router';
import { useMemo } from 'react';
import { SubscanLink } from 'shared/components/widget/SubscanLink';
import { CrossChainStatus } from 'shared/config/constant';
import { HelixHistoryRecord, Network } from 'shared/model';
import { TransferDescription } from './TransferDescription';

export function TargetTx({ record }: { record: HelixHistoryRecord | null }) {
  const { t } = useTranslation();
  const router = useRouter();
  const departure = router.query.from as Network;
  const arrival = router.query.to as Network;

  const content = useMemo(() => {
    if (record?.result === CrossChainStatus.pending) {
      return <Progress percent={50} className="max-w-xs" />;
    }

    if (record?.result === CrossChainStatus.success) {
      return (
        <SubscanLink
          network={arrival}
          txHash={record?.targetTxHash}
          className="hover:opacity-80 transition-opacity duration-200"
        >
          <Typography.Text copyable className="truncate">
            {record?.targetTxHash}
          </Typography.Text>
        </SubscanLink>
      );
    }

    if (record?.result === CrossChainStatus.reverted) {
      return (
        <SubscanLink
          network={departure}
          txHash={record.responseTxHash!}
          className="hover:opacity-80 transition-opacity duration-200"
        >
          <Typography.Text copyable className="truncate">
            {record?.responseTxHash}
          </Typography.Text>
        </SubscanLink>
      );
    }

    return null;
  }, [arrival, departure, record]);

  return (
    <TransferDescription
      title={t('Target Tx Hash')}
      tip={t('Unique character string (TxID) assigned to every verified transaction on the Target Chain.')}
    >
      {content}
    </TransferDescription>
  );
}
