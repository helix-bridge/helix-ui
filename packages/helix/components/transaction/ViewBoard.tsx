interface ViewBoardProps {
  title: string;
  count: string | number;
}

export function ViewBoard({ title, count }: ViewBoardProps) {
  return (
    <div className="flex justify-between items-center lg:flex-col lg:gap-4 bg-gray-200 dark:bg-antDark w-full px-4 lg:px-0 py-2 lg:py-4 text-center text-gray-800 dark:text-gray-400">
      <span className="uppercase">{title}</span>
      <span className="text-xl lg:text-4xl">{count}</span>
    </div>
  );
}
