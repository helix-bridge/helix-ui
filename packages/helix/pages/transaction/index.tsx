import { SearchOutlined, SyncOutlined } from '@ant-design/icons';
import { Affix, Button, Input, Spin, Tooltip } from 'antd';
import { formatDistance, getUnixTime } from 'date-fns';
import { useQuery } from 'graphql-hooks';
import { last } from 'lodash';
import { serverSideTranslations } from 'next-i18next/serverSideTranslations';
import { useRouter } from 'next/router';
import { useCallback, useEffect, useMemo, useRef, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { useVirtual } from 'react-virtual';
import { distinctUntilChanged, filter, Subject } from 'rxjs';
import { CrossChainState } from '@helix/shared/components/widget/CrossChainStatus';
import { EllipsisMiddle } from '@helix/shared/components/widget/EllipsisMiddle';
import { Path } from '@helix/shared/config/constant';
import { useAccountStatistic, useDailyStatistic } from '@helix/shared/hooks';
import { Substrate2SubstrateRecord } from '@helix/shared/model';
import {
  convertToDvm,
  fromWei,
  getChainConfigByName,
  getDisplayName,
  getSupportedChains,
  gqlName,
  isValidAddress,
  prettyNumber,
  revertAccount,
} from '@helix/shared/utils';
import request from 'graphql-request';

const S2S_RECORDS = `
  query s2sRecords($first: Int!, $startTime: Int!, $sender: String) {
    s2sRecords(first: $first, start_timestamp: $startTime, sender: $sender) {
      id
      bridge
      fromChain
      fromChainMode
      toChain
      toChainMode
      laneId
      nonce
      requestTxHash
      responseTxHash
      sender
      recipient
      token
      amount
      startTime
      endTime
      result
      fee
    }
  }
`;

const supportedChains = getSupportedChains();

interface ViewBoardProps {
  title: string;
  count: string | number;
}

function ViewBoard({ title, count }: ViewBoardProps) {
  return (
    <div className="flex justify-between items-center lg:flex-col lg:gap-4 bg-gray-200 dark:bg-antDark w-full px-4 lg:px-0 py-2 lg:py-4 text-center text-gray-800 dark:text-gray-400">
      <span className="uppercase">{title}</span>
      <span className="text-xl lg:text-4xl">{count}</span>
    </div>
  );
}

function Record({ record }: { record: Substrate2SubstrateRecord }) {
  const { fromChainMode, fromChain, sender, recipient, toChain, toChainMode } = record;
  const fromConfig = getChainConfigByName(fromChain, fromChainMode);
  const toConfig = getChainConfigByName(toChain, toChainMode);
  const now = new Date().toISOString().split('.')[0];
  const fromAccount = revertAccount(sender, fromChain, fromChainMode);
  const toAccount = revertAccount(recipient, toChain, toChainMode);
  const amount = fromWei({ value: record.amount, unit: 'gwei' }, prettyNumber);

  return (
    <>
      <span className="justify-self-start ellipse-two-lines">
        {formatDistance(new Date(record.startTime), new Date(now), {
          includeSeconds: true,
          addSuffix: true,
        })}
      </span>

      <div className="flex flex-col col-span-3 overflow-hidden pl-4 pr-2">
        <Tooltip title={fromAccount}>
          <span className="capitalize">{getDisplayName(fromConfig)}</span>
        </Tooltip>
        <EllipsisMiddle isGrow>{fromAccount}</EllipsisMiddle>
      </div>

      <div className="flex flex-col col-span-3 overflow-hidden pl-2 pr-4">
        <Tooltip title={toAccount}>
          <span className="capitalize">{getDisplayName(toConfig)}</span>
        </Tooltip>
        <EllipsisMiddle isGrow>{toAccount}</EllipsisMiddle>
      </div>

      <span>{`${fromChainMode === 'dvm' ? 'x' : ''}${fromConfig?.isTest ? 'O' : ''}RING`}</span>

      <Tooltip title={amount}>
        <span className="justify-self-center max-w-full truncate">{amount}</span>
      </Tooltip>

      <span className="justify-self-center">
        {fromWei({ value: record.fee, unit: fromChainMode === 'dvm' ? 'ether' : 'gwei' })}
      </span>

      <span className="justify-self-center capitalize">{record.bridge}</span>
      <CrossChainState value={record.result} />
    </>
  );
}

const PAGE_SIZE = 50;

function Page({ records }: { records: Substrate2SubstrateRecord[] }) {
  const { t } = useTranslation('common');
  const [isValidSender, setIsValidSender] = useState(true);
  const startTime = useMemo(() => getUnixTime(new Date()), []);
  const router = useRouter();
  const { data: dailyStatistic } = useDailyStatistic();
  const { total: accountTotal } = useAccountStatistic();

  const transactionsTotal = useMemo(
    () => (dailyStatistic?.dailyStatistics || []).reduce((acc, cur) => acc + cur.dailyCount, 0) ?? '-',
    [dailyStatistic]
  );

  const { data, loading, refetch } = useQuery<{ s2sRecords: Substrate2SubstrateRecord[] }>(S2S_RECORDS, {
    variables: { first: PAGE_SIZE, startTime },
  });

  const subject = useMemo(() => new Subject<string>(), []);
  const [source, setSource] = useState<Substrate2SubstrateRecord[]>(records);
  const virtualBoxRef = useRef(null);

  const rowVirtualizer = useVirtual({
    size: source.length + 1,
    parentRef: virtualBoxRef,
    // eslint-disable-next-line no-magic-numbers
    estimateSize: useCallback(() => 60, []),
  });

  useEffect(() => {
    if (data?.s2sRecords && data.s2sRecords.length) {
      setSource((pre) => [...pre, ...data.s2sRecords]);
    }
  }, [data]);

  // eslint-disable-next-line complexity
  useEffect(() => {
    const [lastItem] = [...rowVirtualizer.virtualItems].reverse();
    const courser = last(source);

    if (lastItem && lastItem.index && !loading && courser && lastItem.index >= source.length - 1) {
      subject.next(courser.startTime);
    }
  }, [loading, rowVirtualizer.virtualItems, source, subject]);

  useEffect(() => {
    const sub$$ = subject
      .pipe(
        filter((time) => !!time),
        distinctUntilChanged()
      )
      .subscribe((time: string) => {
        refetch({ variables: { first: PAGE_SIZE, startTime: getUnixTime(new Date(time)) } });
      });

    return () => sub$$.unsubscribe();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return (
    <div>
      <div className="grid lg:grid-cols-3 gap-0 lg:gap-6 place-items-center my-4 lg:my-6">
        <ViewBoard title={t('transactions')} count={transactionsTotal} />
        <ViewBoard title={t('unique users')} count={accountTotal} />
        <ViewBoard title={t('supported blockchains')} count={supportedChains.length} />
      </div>

      <Affix offsetTop={64}>
        <div>
          <div className="flex justify-between">
            <Input
              size="large"
              suffix={<SearchOutlined />}
              allowClear
              onChange={(event) => {
                const value = event.target.value;

                if (!value) {
                  setIsValidSender(true);
                  setSource([]);
                  refetch({ variables: { first: PAGE_SIZE, startTime: getUnixTime(new Date()) } });
                  return;
                }

                try {
                  const address = isValidAddress(value, 'ethereum') ? value : convertToDvm(value);

                  setSource([]);
                  refetch({ variables: { first: PAGE_SIZE, startTime: getUnixTime(new Date()), sender: address } });
                  setIsValidSender(true);
                } catch {
                  setIsValidSender(false);
                }
              }}
              placeholder={t('Search by sender address')}
              className={`max-w-md ${isValidSender ? '' : 'border-red-400'}`}
            />

            <Button
              type="link"
              onClick={() => {
                setSource([]);
                refetch({ variables: { first: PAGE_SIZE, startTime: getUnixTime(new Date()) } });
              }}
              disabled={loading}
              className="flex items-center cursor-pointer"
            >
              <span className="mr-2">{t('Latest transactions')}</span>
              <SyncOutlined />
            </Button>
          </div>

          <div className="mt-4 lg:mt-6 grid grid-cols-12 border-b border-gray-500 items-center p-4 bg-gray-200 dark:bg-antDark">
            <span className="mr-5">{t('Time')}</span>
            <span className="col-span-3 overflow-hidden pl-4 pr-2">{t('From')}</span>
            <span className="col-span-3 overflow-hidden pl-2 pr-4">{t('To')}</span>
            <span>{t('Asset')}</span>
            <span className="justify-self-center">{t('Amount')}</span>
            <span className="justify-self-center">{t('Fee')}</span>
            <span className="justify-self-center">{t('Bridge')}</span>
            <span>{t('Status')}</span>
          </div>

          <div
            // header: 64px search box: 40px table header + margin: 78px footer 54px
            style={{ height: 'calc(100vh - 40px - 64px - 78px - 54px)' }}
            ref={virtualBoxRef}
            className="overflow-auto"
          >
            {data && (
              <div style={{ height: rowVirtualizer.totalSize, width: '100%', position: 'relative' }}>
                {rowVirtualizer.virtualItems.map((row) => {
                  const record = source[row.index];
                  const isLoaderRow = row.index > source?.length - 1;
                  // eslint-disable-next-line no-magic-numbers
                  const bg = row.index % 2 === 0 ? 'bg-gray-200 dark:bg-antDark' : '';

                  return (
                    <div
                      className={`grid grid-cols-12 items-center py-2 px-4 cursor-pointer transition-all duration-300 
                      hover:bg-gray-400 dark:hover:bg-gray-800 ${bg}`}
                      style={{
                        position: 'absolute',
                        top: 0,
                        left: 0,
                        width: '100%',
                        height: `${row.size}px`,
                        transform: `translateY(${row.start}px)`,
                      }}
                      key={row.key}
                      onClick={() => {
                        router.push({
                          pathname: Path.transaction + '/' + record.id,
                          query: new URLSearchParams({
                            from: record.fromChain,
                            to: record.toChain,
                            fromMode: record.fromChainMode,
                            toMode: record.toChainMode,
                          }).toString(),
                        });
                      }}
                    >
                      {isLoaderRow ? (
                        data.s2sRecords.length ? (
                          <Spin className="col-span-full" />
                        ) : (
                          <span className="col-span-full text-center">{t('Nothing more to load')}</span>
                        )
                      ) : (
                        record && <Record record={record} />
                      )}
                    </div>
                  );
                })}
              </div>
            )}
          </div>
        </div>
      </Affix>
    </div>
  );
}

export const getServerSideProps = async ({ locale }: { locale: string }) => {
  const translations = await serverSideTranslations(locale ?? 'en', ['common']);
  const startTime = getUnixTime(new Date());
  const url =
    process.env.NODE_ENV === 'development'
      ? 'http://localhost:4002/graphql'
      : 'https://wormhole-apollo.darwinia.network/';
  const records = await request(url, S2S_RECORDS, {
    first: PAGE_SIZE,
    startTime,
  }).then((res) => res && res[gqlName(S2S_RECORDS)]);

  return {
    props: {
      ...translations,
      records,
    },
  };
};

export default Page;
