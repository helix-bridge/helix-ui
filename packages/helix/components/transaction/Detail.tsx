import { Divider, Typography } from 'antd';
import { useRouter } from 'next/router';
import { useMemo } from 'react';
import { useITranslation } from 'shared/hooks';
import { HelixHistoryRecord, Network } from 'shared/model';
import { isCrabDVMHeco } from 'shared/utils/bridge';
import { fromWei, prettyNumber, revertAccount } from 'shared/utils/helper';
import { getChainConfig } from 'shared/utils/network';
import { TransferStep } from '../../model/transfer';
import { IBreadcrumb } from './Breadcrumb';
import { Bridge } from './Bridge';
import { SourceTx } from './SourceTx';
import { TargetTx } from './TargetTx';
import { Timestamp } from './Timestamp';
import { TransferDescription } from './TransferDescription';
import { TransferDetail } from './TransferDetail';
import { TxStatus } from './TxStatus';

interface DetailProps {
  record: HelixHistoryRecord;
  transfers: TransferStep[];
}

export function Detail({ record, transfers }: DetailProps) {
  const { t } = useITranslation();
  const router = useRouter();
  const departure = getChainConfig(router.query.from as Network);
  const arrival = getChainConfig(router.query.to as Network);

  const amount = useMemo(() => {
    const token = departure.tokens.find((item) => item.symbol.toLowerCase() === record.token.toLowerCase());

    return fromWei({ value: record?.amount ?? 0, decimals: token?.decimals ?? 9 }, prettyNumber);
  }, [departure.tokens, record?.amount, record.token]);

  const feeDecimals = useMemo(() => {
    const token = departure.tokens.find((item) => item.symbol.toLowerCase() === record.feeToken.toLowerCase());

    return token?.decimals ?? 9;
  }, [departure.tokens, record.feeToken]);

  return (
    <>
      <IBreadcrumb txHash={record.requestTxHash} />

      <div className="flex justify-between items-center mt-6">
        <h3 className="uppercase text-xs md:text-lg">{t('transaction detail')}</h3>
        <Bridge />
      </div>

      <div className="px-8 py-3 mt-6 bg-gray-200 dark:bg-antDark">
        <SourceTx hash={record.requestTxHash} />

        <TargetTx record={record} />

        <TxStatus record={record} />

        <Timestamp record={record} />

        <Divider />

        <TransferDescription title={t('Sender')} tip={t('Address (external or contract) sending the transaction.')}>
          <Typography.Text copyable className="truncate">
            {revertAccount(record.sender, departure)}
          </Typography.Text>
        </TransferDescription>

        <TransferDescription title={t('Receiver')} tip={t('Address (external or contract) receiving the transaction.')}>
          <Typography.Text copyable className="truncate">
            {revertAccount(record.recipient, arrival)}
          </Typography.Text>
        </TransferDescription>

        {!!transfers.length && <TransferDetail transfers={transfers} amount={amount} />}

        <Divider />

        <TransferDescription
          title={t('Value')}
          tip={t('The amount to be transferred to the recipient with the cross-chain transaction.')}
        >
          {amount} {transfers[0].token.name}
        </TransferDescription>

        <TransferDescription
          title={t('Transaction Fee')}
          tip={'Amount paid for processing the cross-chain transaction.'}
        >
          {fromWei({ value: record.fee, decimals: feeDecimals })} {record.feeToken === 'null' ? null : record.feeToken}
        </TransferDescription>

        <Divider />

        <TransferDescription title={t('Nonce')} tip={t('A unique number of cross-chain transaction in Bridge')}>
          {isCrabDVMHeco(record.fromChain, record.toChain) ? record.laneId : record.nonce}
        </TransferDescription>
      </div>
    </>
  );
}
