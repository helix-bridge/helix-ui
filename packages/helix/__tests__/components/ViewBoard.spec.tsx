/// <reference types="jest" />

import { create } from 'react-test-renderer';
import { ViewBoard } from '../../components/transaction/ViewBoard';

describe('<ViewBoard />', () => {
  it('render history item', () => {
    const component = create(<ViewBoard title="transaction" count={90} />);

    let tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });
});
