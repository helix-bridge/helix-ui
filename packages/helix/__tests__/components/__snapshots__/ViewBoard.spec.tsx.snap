// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ViewBoard /> render history item 1`] = `
<div
  className="flex justify-between items-center lg:flex-col lg:gap-4 bg-gray-200 dark:bg-antDark w-full px-4 lg:px-0 py-2 lg:py-4 text-center text-gray-800 dark:text-gray-400"
>
  <span
    className="uppercase"
  >
    transaction
  </span>
  <span
    className="text-xl lg:text-4xl text-white"
  >
    90
  </span>
</div>
`;
